### PR TITLE
Create SQL Grader XBlock

### PR DESCRIPTION
**Pull request by stvstnfrd**

### CODE REVIEW

File: sql_grader/datasets/rating.sql [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (stvstnfrd) [inline sql_grader/datasets/rating.sql:1]: These two `.sql` files in `datasets/` subdirectory are used to see a database.

File: sql_grader/mixins/scenario.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class XBlockWorkbenchMixin(object):
+    """
+    Provide a default test workbench for the XBlock
+    """
+    # pylint: disable=too-few-public-methods
+
+    @classmethod
+    def workbench_scenarios(cls):
```
Comment (stvstnfrd) [inline sql_grader/mixins/scenario.py:60]: Autodetect scenarios as `.xml` files in the `scenarios` subdirectory.

File: sql_grader/models.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    # pylint: disable=too-few-public-methods
+
+    display_name = String(
+        display_name=_('Display Name'),
+        help=_('BLAH BLAH BE CREATIVE'),
+        default=_('SQL Problem'),
+        scope=Scope.content,
+    )
+    # TODO: Make this a choice from a list
+    dataset = String(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: Seed the database with one of the `.sql` files initially.

File: sql_grader/models.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        display_name=_('Dataset'),
+        help=_('Which initial dataset/database to be used for queries'),
+        default='rating',
+        scope=Scope.content,
+    )
+    answer_query = String(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: This is a SQL query which produces the correct results. Since this is evaluated by a SQL engine, there can be many correct answers, up to and including actually hard-coding results like:

```sql
SELECT 1
UNION ALL
SELECT 2
-- etc.
;
```

File: sql_grader/models.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        display_name=_('Answer Query'),
+        help=_('A correct response SQL query'),
+        default='',
+        scope=Scope.content,
+        multiline_editor=True,
+    )
+    verify_query = String(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: This is used for queries which modify the database (`UPDATE`, `INSERT`, `DELETE`),
so that they can verify the changes made by the `answer_query`.

The idea being a setup like:

```sql
-- answer_query
DELETE * FROM Movies WHERE movieId > 5;
```

```sql
-- verify_query
SELECT * FROM Movie WHERE movieId > 5;
-- this should return zero results
```

This could use better documentation.

File: sql_grader/models.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        display_name=_('Verify Query'),
+        help=_('A verification SQL query'),
+        default='',
+        scope=Scope.content,
+        multiline_editor=True,
+    )
+    answer_result = List(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: This is optional; I'm caching the correct answer to avoid additional invocations of `sqlite`.

File: sql_grader/models.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        display_name=_('Answer Result'),
+        help=_('Results from the correct SQL query'),
+        default=None,
+        scope=Scope.content,
+    )
+    is_ordered = Boolean(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: Whether or not order of results should matter.

File: sql_grader/models.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        'prompt',
+        'weight',
+    ]
+    prompt = String(
+        display_name=_('Prompt'),
+        help=_('Explanatory text to accompany the problem'),
+        default='',
+        scope=Scope.content,
+    )
+    submission_query = String(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: A user's answer.

File: sql_grader/models.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        display_name=_('Submission Query'),
+        help=_('A Submission Query'),
+        default='',
+        scope=Scope.user_state,
+    )
+    submission_result = List(
```
Comment (stvstnfrd) [inline sql_grader/models.py]: The output from a user's answer.
This is optional and could be removed.

File: sql_grader/models.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            'score': self.score,
+            'score_weighted': int(self.score * self.weight),
+            'max_score': int(self.max_score()),
+            'user_alert': '',  # TODO: Remove this?
+            'error_class': error_class,
+        })
+        return context
+
+
+class SqlProblem(object):
```
Comment (stvstnfrd) [inline sql_grader/models.py]: Note: if the use of this class seems unorthodox for an XBlock, that's because I originally wrote this in pure Python, prior to converting to an XBlock. I'm open to refactoring, if it makes sense.

File: sql_grader/models.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        self.is_ordered = is_ordered
+        self.answer_query = answer_query
+        self.verify_query = verify_query
+        self.answer_result, _ = SqlProblem.run_query(
+            database,
+            answer_query,
+            verify_query,
+        )
+
+    def attempt(self, query):
```
Comment (stvstnfrd) [inline sql_grader/models.py:46]: Attempt to answer the problem with the provided query.

File: sql_grader/models.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        This only needs run once, at startup.
+        """
+        with open(path_sql, 'r') as sql:
+            query = sql.read()
+        connection = cls.create_database_from_sql(query)
+        return connection
+
+    @classmethod
+    def create_database_from_sql(cls, query):
+        connection = sqlite3.connect(':memory:', check_same_thread=False)
```
Comment (stvstnfrd) [inline sql_grader/models.py:79]: Now we're getting to the fun stuff!

Basically, create a new in-memory database with SQLite and then execute a query against this new DB.

File: sql_grader/models.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        return connection
+
+    @staticmethod
+    def clone_database(source):
+        """
+        Copy the contents of a source database into a new in-memory database
+
+        This should be run for each request.
+        """
+        destination = sqlite3.connect(':memory:', check_same_thread=False)
```
Comment (stvstnfrd) [inline sql_grader/models.py:91]: Create another in-memory DB and seed it with the output from another DB.

File: sql_grader/models.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                    for row in connection.execute(query):
+                        # print('row', row)
+                        result.append(row)
+                except Exception as error:
+                    print('ERROR', error)
+                    result = None
+                    message = unicode(error)
+                    # TODO: more robust error handling
+            return result, message
+        database = cls.clone_database(source)
```
Comment (stvstnfrd) [inline sql_grader/models.py:112]: Note: We always run the user query against a new copy of the in-memory DB.

File: sql_grader/models.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+        # print('*' * 20)
+        result, error = run(database, query)
+        if verify_query:
```
Comment (stvstnfrd) [inline sql_grader/models.py:114]: Queries that modify the state of the DB require an additional verification step.

File: sql_grader/models.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+            # print('verify_query', verify_query)
+            result, _ = run(database, verify_query)
+        return result, error
+
+    @staticmethod
+    def compare_rows(expected, actual, is_ordered=True):
+        expected = expected or []
+        actual = actual or []
+        if len(expected) != len(actual):
```
Comment (stvstnfrd) [inline sql_grader/models.py:125]: Short circuit quickly if the counts don't match, without bothering to compare rows.

File: sql_grader/models.py [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            return False
+        if not is_ordered:
+            expected = sorted(expected)
+            actual = sorted(actual)
```
Comment (stvstnfrd) [inline sql_grader/models.py:129]: If we don't care about order, go ahead and sort the `actual` and `expected`.

File: sql_grader/scenarios/SQL-Labs.xml [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+            display_name="Q1"
+            answer_query="SELECT title FROM Movie where Director = 'Steven Spielberg';"
+            is_ordered="False"
+            prompt="Find the titles of all movies directed by Steven Spielberg."
+            weight="0"
+        />
+        <sql_grader
+            dataset="rating"
+            display_name="Q2"
+            answer_query="SELECT title FROM Movie where Director = 'Steven Spielberg';"
```
Comment (stvstnfrd) [inline sql_grader/scenarios/SQL-Labs.xml:14]: TODO: My original plan was to implement all of the real course problems inside the scenario :) The prompts here are the actual real prompts from the course; the answers just need updated.

File: sql_grader/scenarios/SQL-Labs.xml [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+                    View the database
+                </a>.
+                (You can also
+                <a target="_blank" href="https://s3-us-west-2.amazonaws.com/prod-c2g/db/Winter2013/files/rating.sql">
+                    download the schema and data
+                </a>
+                .)
+            </p>
+            <p>
+                <strong>Instructions: </strong>
```
Comment (stvstnfrd) [inline sql_grader/scenarios/SQL-Labs.xml:117]: Note: this is all existing messaging from the current course.

File: sql_grader/static/view.css [hunk 20]
```diff
@@ -0,0 +1,41 @@
+.sql_grader_block {
+  /* WORKBENCH ONLY */
```
Comment (stvstnfrd) [inline sql_grader/static/view.css]: TODO: Cleanup

File: sql_grader/static/view.js [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    var options = {
+        value: $textarea.val(),
+        mode: 'text/x-sql',
+        indentUnit: 4,
+        lineNumbers: true,
+        indentWithTabs: true,
+        smartIndent: true,
+        matchBrackets : true,
+        extraKeys: {"Ctrl-Space": "autocomplete"},
+        /*hintOptions: {
```
Comment (stvstnfrd) [inline sql_grader/static/view.js]: TODO: Get this working or remove it.

File: sql_grader/xblocks.py [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+def resource_string(path):
+    """
+    Handy helper for getting resources from our kit
+    """
+    data = pkg_resources.resource_string(__name__, path)
+    return data.decode('utf8')
+
+
+def all_datasets():
```
Comment (stvstnfrd) [inline sql_grader/xblocks.py]: Dynamically load all available datasets from disk (any `.sql` file in the `datasets` subdirectory.
This can be extended by operators at install/deploy time.

File: sql_grader/xblocks.py [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    dataset_directory = pkg_resources.resource_filename('sql_grader', 'datasets')
+    for dirname, _, files in os.walk(dataset_directory):
+        for fname in files:
+            if fname.endswith('.sql'):
+                fname = fname[:-4]
+                yield fname
+
+
+def create_database(database_name):
```
Comment (stvstnfrd) [inline sql_grader/xblocks.py]: Create an in-memory database from a discovered `.sql` dataset.

File: sql_grader/xblocks.py [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+    pathname = 'datasets/' + database_name + '.sql'
+    contents = resource_string(pathname)
+    # print(contents)
+    database = SqlProblem.create_database_from_sql(contents)
+    return database
+
+
+DATABASES = {
```
Comment (stvstnfrd) [inline sql_grader/xblocks.py]: Load a copy of all databases into memory at startup time, to avoid overhead and disk access on each attempt/load.

We'll clone these on a per-request basis.

**Review by stvstnfrd — COMMENTED** [review]
@ormsbee Here's my first take.

I notice there are still a few print/debug/comment statements that can be removed;
I'll do a cleanup pass.

Also, I'm missing docstrings in a few places, so I'll add those too.

It's been a while since I first built this, so please ask for any clarification that's helpful :)

File: Dockerfile [hunk 25]
```diff
@@ -0,0 +1,8 @@
+FROM stvstnfrd/workbench:latest
+MAINTAINER stv <stv@stanford.edu>
```
Comment (ormsbee) [inline Dockerfile]: Please update ;-)

File: sql_grader/mixins/dates.py [hunk 26]
```diff
@@ -0,0 +1,33 @@
+"""
+Mixin date-related functionality
+"""
+import datetime
+
+
+class EnforceDueDates(object):  # pylint: disable=too-few-public-methods
```
Comment (ormsbee) [inline sql_grader/mixins/dates.py]: Nit: Python 2 is gone, so you don't need to inherit from `object` any longer.

File: sql_grader/mixins/dates.py [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+    """
+
+    def is_past_due(self):
+        """
+        Determine if component is past-due
+        """
+        # These values are pulled from platform.
+        # They are defaulted to None for tests.
+        due = getattr(self, 'due', None)
+        graceperiod = getattr(self, 'graceperiod', None)
```
Comment (ormsbee) [inline sql_grader/mixins/dates.py:21]: Nit: I think you can declare these as actual fields in the XBlock and it will get the inherited value if it exists. I don't think you need to do the `getattr` bit, though this should still work.

File: sql_grader/mixins/fragment.py [hunk 28]
```diff
@@ -0,0 +1,94 @@
+"""
+Mixin fragment/html behavior into XBlocks
+
+Note: We should resume test coverage for all lines in this file once
+split into its own library.
+"""
+from __future__ import absolute_import
+
+from xblock.core import XBlock
+from xblock.fragment import Fragment
```
Comment (ormsbee) [inline sql_grader/mixins/fragment.py]: As per the comments in https://github.com/edx/XBlock/blob/master/xblock/fragment.py , this has been deprecated (moved to a separately `web_fragments` lib, so `web_fragments.fragment.Fragment`).

File: sql_grader/mixins/grading.py [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def _(text):
+    """
+    Mock translation for scraping
+    """
+    return text
+
+
+class Scorable(object):
```
Comment (ormsbee) [inline sql_grader/mixins/grading.py]: There's a dedicated mixin for this in the XBlock repo now: https://github.com/edx/XBlock/blob/master/xblock/scorable.py

File: sql_grader/models.py [hunk 30]
```diff
@@ -0,0 +1,222 @@
...
+from xblock.fields import Scope
+from xblock.fields import String
+
+
+class XBlockDataMixin(object):
+    # pylint: disable=too-few-public-methods
+
+    display_name = String(
+        display_name=_('Display Name'),
+        help=_('BLAH BLAH BE CREATIVE'),
```
Comment (ormsbee) [inline sql_grader/models.py]: The generic one for this is: "The display name for this component." Not saying it's brilliant and helpful, but we have translations for it at least.

File: sql_grader/models.py (hunk 5 above)
Comment (ormsbee) [reply sql_grader/models.py]: Yeah, adding a bit to the "help" section for this would be great.

File: sql_grader/models.py (hunk 9 above)
Comment (ormsbee) [reply sql_grader/models.py]: I would mostly be concerned by how large this could get. If it can be hard-capped at something reasonable, it seems fine to have. Or we can remove to make simpler–I don't have a strong opinion.

File: sql_grader/models.py (hunk 10 above)
Comment (ormsbee) [reply sql_grader/models.py]: I think it's fine to have that separation (preferable, even), but please add a docstring to say what the purpose of this class is.

File: sql_grader/models.py [hunk 31]
```diff
@@ -0,0 +1,222 @@
...
+    # block-scope
+    dataset = None
+    answer_query = None
+    verify_query = None
+    answer_result = None
+    is_ordered = True
+
+    # user-scope
+    submission_query = None
+    submission_result = None
```
Comment (ormsbee) [inline sql_grader/models.py]: Why are these class attributes? Wouldn't that have side-effect issues across requests?

**Review by ormsbee — CHANGES_REQUESTED** [review]
WIP review. Going to step away for a bit so wanted to get this to you earlier rather than later. Will review the rest later this afternoon.

File: sql_grader/models.py (hunk 14 above)
Comment (ormsbee) [reply sql_grader/models.py:112]: So it's great that every student gets their own in-memory database, and we don't have to worry about SQL injection. But do we have to worry about any kind of escaping out of sqlite3? Does sqlite3 allow you to issue non-SQL commands that say, read something from disk and echo it somewhere?

File: sql_grader/static/codemirror/lib/codemirror.js [hunk 32]
```diff
@@ -0,0 +1,9693 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
```
Comment (ormsbee) [inline sql_grader/static/codemirror/lib/codemirror.js:1]: This won't conflict with existing CodeMirror code in platform, right?

File: sql_grader/datasets/rating.sql [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+
+/* Populate the tables with our data */
+insert into Movie values(101, 'Gone with the Wind', 1939, 'Victor Fleming');
+insert into Movie values(102, 'Star Wars', 1977, 'George Lucas');
+insert into Movie values(103, 'The Sound of Music', 1965, 'Robert Wise');
+insert into Movie values(104, 'E.T.', 1982, 'Steven Spielberg');
+insert into Movie values(105, 'Titanic', 1997, 'James Cameron');
+insert into Movie values(106, 'Snow White', 1937, null);
+insert into Movie values(107, 'Avatar', 2009, 'James Cameron');
+insert into Movie values(108, 'Raiders of the Lost Ark', 1981, 'Steven Spielberg');
```
Comment (ormsbee) [inline sql_grader/datasets/rating.sql:23]: FWIW, if you run into performance issues (which seems unlikely), you might want to try reducing the number of SQL statements so you have one "insert" per table, and then a bunch of values, e.g.:

```
insert into Movie (mID, title, year, director)
    values
        (101, 'Gone with the Wind', 1939, 'Victor Fleming'),
        (102, 'Star Wars', 1977, 'George Lucas'),
        ...
```

File: sql_grader/xblocks.py (hunk 24 above)
Comment (ormsbee) [reply sql_grader/xblocks.py]: What is this overhead, btw? FWIW, if it's small enough where it wouldn't be noticed, I'd be inclined to keep it simpler, recompute each time, and not have thread-level persistent state. If it's significantly larger, I'd prefer to keep it out of the module level init so that it doesn't factor into startup time for Studio/LMS (which is already seconds long–we end up doing an audit every year or so to cut stuff out). If we're talking a few tens of ms, please just do it in the simplest, least-dangerous way possible (i.e. do all the copying in the handler with no caching).

File: sql_grader/mixins/dates.py (hunk 26 above)
Comment (stvstnfrd) [reply sql_grader/mixins/dates.py]: Done in pending commit.

Updated all relevant classes.

File: sql_grader/mixins/fragment.py (hunk 28 above)
Comment (stvstnfrd) [reply sql_grader/mixins/fragment.py]: Done in pending commit.

File: sql_grader/models.py (hunk 30 above)
Comment (stvstnfrd) [reply sql_grader/models.py]: Done in pending commit.

File: sql_grader/datasets/rating.sql (hunk 33 above)
Comment (stvstnfrd) [reply sql_grader/datasets/rating.sql:23]: Okay, yeah, I pulled these `.sql` files directly from the course's files & uploads :)

File: sql_grader/mixins/dates.py (hunk 27 above)
Comment (stvstnfrd) [reply sql_grader/mixins/dates.py:21]: Ah, yeah, this is me cargo-culting myself from old XBlock code from years ago.

TODO: I'll try refactoring as such.

File: sql_grader/mixins/grading.py (hunk 29 above)
Comment (stvstnfrd) [reply sql_grader/mixins/grading.py]: TODO: I'll refactor to use this...

File: sql_grader/models.py (hunk 9 above)
Comment (stvstnfrd) [reply sql_grader/models.py]: Looking at it now, I think this was either unnecessary future-proofing or a relic of my conversion from pure Python.

We don't appear to be using this for anything worthwhile and, since we're saving their actual submission (this variable here actually represents the output from their SQL statement), we can always re-calculate `submission_result`.

I'll drop this instance member entirely for now.

Done in pending commit.

File: sql_grader/models.py (hunk 14 above)
Comment (stvstnfrd) [reply sql_grader/models.py:112]: # TL;DR: I haven't been able to find any such explicit capability.

The closest thing I can find is `sqlite`'s special `dot-commands` [1], which do allow for some related functionality (change the working directory, at least). But that is handled exclusively by the `sqlite` _executable_ and _not_ by the `sqlite` _library_.
I did confirm that I could _not_ issue dot-commands via the Python implementation of `executescript` (which is the one integration point where we pass user input) [2].

From the `sqlite` docs [8]:

## SQLite Always Validates Its Inputs

>  SQLite should never crash, overflow a buffer, leak memory, or exhibit any other harmful behavior, even when presented with maliciously malformed SQL inputs or database files. SQLite should always detect erroneous inputs and raise an error, not crash or corrupt memory. Any malfunction caused by an SQL input or database file is considered a serious bug and will be promptly addressed when brought to the attention of the SQLite developers. SQLite is extensively fuzz-tested to help ensure that it is resistant to these kinds of errors.

> Nevertheless, bugs happen. If you are writing an application that sends untrusted SQL inputs or database files to SQLite, there are additional steps you can take to help reduce the attack surface and prevent zero-day exploits caused by undetected bugs. 

It appears that these settings may only be available as compile-time options [9], so I'm not sure which values are used in our Python deploy and/or the feasibility of modifying them.

Perhaps just setting some reasonable max length (could be defaulted and possibly overridden in Studio) accomplishes the 80/20 for us?

That said, with full disclosure, there do appear to have been a few `sqlite` CVEs throughout the years related to RCE [3][4][5][6][7], though it's not clear whether or not they would apply in our use cases (or are any "different" than CVEs detected in other libraries we use (Django, oauth, etc) [10][11]). A common attack mode revolves around first inserting "attack code" into a persistent database, which is then later read out by the unsuspecting application. Others involve injection. Those such cases shouldn't impact us, given we're only using write-once, in-memory databases.

It's possible there are relevant attack vectors I'm not seeing, so I'd appreciate a second set of eyes (and thumbs up!) on this, before proceeding ;)

Thinking of contingencies if we're still concerned about sandboxing this, how does the Code Jail model differ and would it be worth considering an implementation of a Custom Python Grader. It'd be a much less elegant solution, but it seems feasible?

Anyway, take a look, think it over, and let me know if you want to discuss any more of this further in depth.

Thanks!

- [1] https://sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_
- [2] https://docs.python.org/3.5/library/sqlite3.html#sqlite3.Cursor.executescript
- [3] https://nakedsecurity.sophos.com/2018/12/19/sqlite-creator-fires-back-at-tencents-bug-hunters/
- [4] https://research.checkpoint.com/2019/select-code_execution-from-using-sqlite/
- [5] https://i.blackhat.com/USA-19/Thursday/us-19-Qian-Exploring-The-New-World-Remote-Exploitation-Of-SQLite-And-Curl-wp.pdf
- [6] https://www.blackhat.com/docs/us-17/wednesday/us-17-Feng-Many-Birds-One-Stone-Exploiting-A-Single-SQLite-Vulnerability-Across-Multiple-Software.pdf
- [7] https://www.theregister.co.uk/2019/05/10/sqlite_rce_vuln/
- [8] https://www.sqlite.org/security.html
- [9] https://www.sqlite.org/security.html#untrusted_sql_inputs
- [10] https://www.cvedetails.com/product/16355/Sqlite-Sqlite.html?vendor_id=9237
- [11] https://www.cvedetails.com/product/18211/Djangoproject-Django.html?vendor_id=10199

File: Dockerfile (hunk 25 above)
Comment (stvstnfrd) [reply Dockerfile]: Removed in pending commit.

File: sql_grader/models.py (hunk 10 above)
Comment (stvstnfrd) [reply sql_grader/models.py]: Done.

File: sql_grader/models.py (hunk 31 above)
Comment (stvstnfrd) [reply sql_grader/models.py]: Design relics that predate the XBlock conversion and optimization.

These are removed in a pending commit.

File: sql_grader/xblocks.py (hunk 24 above)
Comment (stvstnfrd) [reply sql_grader/xblocks.py]: The overhead is disk I/O.

I haven't run explicit benchmarks, but could.

Instinctively, I wanted to avoid the disk access on each page load/submit, as a best practice.

I've factored out the two user-scoped fields from the class (they were design relics), so I'm not too concerned about state persistence now; these objects are effectively read-only and used only to clone each per-load/submit.

That said, I can still refactor, if you'd prefer ;)

File: sql_grader/models.py (hunk 5 above)
Comment (stvstnfrd) [reply sql_grader/models.py]: Done.

**Review by stvstnfrd — COMMENTED** [review]
Most changes have already been made and are being pushed now;
there's still one or two rectors outstanding, so I'll ping when those are ready.

I've added a question in the comments, but otherwise am :+1: with the feedback.

File: sql_grader/static/codemirror/lib/codemirror.js (hunk 32 above)
Comment (stvstnfrd) [reply sql_grader/static/codemirror/lib/codemirror.js:1]: TODO: Confirm Studio integration.

File: sql_grader/xblocks.py (hunk 24 above)
Comment (ormsbee) [reply sql_grader/xblocks.py]: Yeah, let's not worry about it for now.

Comment (ormsbee) [reply sql_grader/xblocks.py]: As in, let's keep it as-is and come back to it if it turns out to be a problem.

File: sql_grader/mixins/dates.py (hunk 27 above)
Comment (stvstnfrd) [reply sql_grader/mixins/dates.py:21]: Omitting for now.

**Review by stvstnfrd — COMMENTED** [review]
@ormsbee Refactoring done and documentation updated.

Relevant commit here: https://github.com/Stanford-Online/xblock-sql-grader/pull/1/commits/3970ed5c1ba181dad6fd051da8ff48a846acdddc
(note: Debug/comments already cleaned up in a separate commit.)

**Review by ormsbee — APPROVED** [review]
